sql: introduce virtual sequences and use with SERIAL

knz opened the pull request:
Very good idea by @BramGruneir 

New features: virtual sequences (opt-in) and create-sequence-upon-SERIAL (opt-in). See individual commits for details and release notes.

For reference:

- I made everything opt-in here (i.e. default to previous CockroachDB behavior) to minimize risk.
- UX with Hibernate will thus not automatically be improved. We need to inform users (and upgrade our tutorials/docs) that this experimental feature is available and how to use it to unblock problems.
- It is possible that making the new default be `virtual_sequence` will both be backward-compatible with performance and unlock UX for Hibernate users. This has @awoods187 preference I think. However I propose to only flip that switch after we've seen a week or two pass with this patch merged.

Fixes #22607 (using the new setting set to `sql_sequence`).
Fixes #24062 (using the new setting set to `sql_sequence`).
Fixes #26730 (using the new setting set to either `virtual_sequence` or `sql_sequence`).

knz committed:
sql: introduce virtual sequences

Suggested/recommended by @BramGruneir.

This patch introduces **virtual sequences**, a CockroachDB-specific
extension to SQL sequences. Virtual sequences are sequences that do
not generate monotonically increasing values and instead produce
values like the built-in function `unique_rowid()`.

They are intended for use in combination with the PostgreSQL column
pseudo-type "SERIAL", where clients expect a sequence object to be
created but usually don't really care about the particular values
generated by the sequence. Virtual sequences buy us a behavior that's
compatible schema-wise with PostgreSQL but without the performance
scalability bottleneck of regular sequences.

Technically, a virtual sequence is supported by a regular sequence
descriptor with just an extra option called "virtual". Therefore it
behaves as a database object much alike a regular SQL sequence and can
be manipulated with the same DDL (CREATE/DROP/SHOW) etc. and thus
appears very much alike a regular SQL sequence to clients -- to ensure
they can be used as a substitute to regular SQL sequences without
clients noticing.

The difference between a virtual sequence and a regular sequence is
only to be found in the behavior of the `nextval()` built-in function:
with virtual sequences this skips the KV increment operation and
instead runs the code of `unique_rowid()`.

Release note (sql change): A CockroachDB-specific experimental
extension to SQL sequences, called "virtual sequences", is introduced
as an experiment to more transparently support PostgreSQL applications
using the SERIAL pseudo-type.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/28575)
<!-- Reviewable:end -->

knz commented:
I tested locally and this seems to work well. Please have a look already. However I still need to iterate to iron out a couple of test failures.

knz commented:
oh it so appears that there are no test failures. Wowza

knz committed:
sql: hoist seq creation from `(*createSequenceNode).startExec()`

This will enable reuse in other places.

Release note: None

knz committed:
sql: make the SERIAL pseudo-type optionally auto-create sequence

Prior to this patch, SERIAL in CockroachDB was always alias for `INT
DEFAULT unique_rowid()`. However 3rd party apps developed for
PostgreSQL expect SERIAL to associate a column with a SQL sequence.

Although the particular values produced for the SERIAL column rarely
matter to the client app, we have found in practice that client
apps (or ORM frameworks) are likely to assert that there is a sequence
object created in the namespace with a particular name, and that they
are able to call `nextval()` / `currval()` on that object while
populating new rows on the table.

This patch increases compatibility with those apps by optionally
associating SERIAL columns to sequences like PostgreSQL.

This behavior is driven by a new session variable called
`experimental_serial_normalization`. Its default for new SQL sessions
is configured via the new cluster setting
`sql.defaults.serial_normalization`. The three values are:

- `rowid` (the default): SERIAL behaves like before and expands to
  `INT NOT NULL DEFAULT unique_rowid()`. The SERIAL
  width (SMALLSERIAL/SERIAL4/etc) is ignored.
- `virtual_sequence`: SERIAL creates a virtual sequence and expands
  to `INT NOT NULL DEFAULT nextval(...seqname...)`. The SERIAL
  width (SMALLSERIAL/SERIAL4/etc) is ignored.
- `sql_sequence`: This is the PostgreSQL "stricter" compatibility
  mode. SERIAL creates a regular SQL sequence and expands to
  `...inttype... NOT NULL DEFAULT nextval(...seqname...)`. The
  SERIAL width (SMALLSERIAL/SERIAL4/etc) is respected.

Finally, this patch ensures that SERIAL always implies NOT NULL, as
required for PostgreSQL compatibility, even when SERIAL is not used as
PRIMARY KEY.

Release note (sql change): CockroachDB now supports two experimental
compatibility modes with how PostgreSQL handles SERIAL and sequences,
to ease reuse of 3rd party frameworks or apps developed for
PostgreSQL. These modes can be enabled with the session variable
`experimental_serial_normalization` (per client) and cluster setting
`sql.defaults.serial_normalization` (cluster-wide). The first mode
`virtual_sequence` enables compatibility with many applications using
SERIAL with maximum performance and scalability. The second mode
`sql_sequence` enables maximum PostgreSQL compatibility but thus uses
regular SQL sequences and is thus subject to performance constraints.

knz commented:
@mjibson since you plan to look at this let me flesh out the configuration options that andy W, Bram and I have discussed.

There would be a new cluster setting (name TBD) to serve as cluster-wide default.
Then there would be a new session var (name TBD) which, in every session, defaults to the cluster setting. This is the same configuration mechanism we're using for `distsql` and `experimental_opt` already.

The setting would have the following 3 values:

- `rowid`: SERIAL maps to `INT NOT NULL DEFAULT unique_rowid()`
- `virtual_sequence`: SERIAL maps to `INT NOT NULL DEFAULT nextval(seqname)` with a sequence created as per `CREATE SEQUENCE seqname  VIRTUAL`
- `sql_sequence`: SERIAL maps to sequences like PostgreSQL.

knz commented:
We're thinking about having the cluster setting default to `virtual_sequence` for *new* clusters, `rowid` for *existing* clusters, with `sql_sequence` become opt-in for client apps that are still not happy with virtual sequences.

mjibson commented:
Reading over the code, it looks like these new virtual sequences are a sugar wrapper over unique_rowid. The IMPORT stuff calls that same function, so I think the actual bytes generated by IMPORT are compatible with this. It appears that IMPORT just needs to be taught about SERIAL or something. Overall I think this will be an easy fix for IMPORT.

mjibson commented:
I think we have two options here. The first and easiest one is to have IMPORT convert SERIAL like the `rowid` option you describe above to avoid the sequence descriptor problem completely. I'm completely ok with that for now.

The second option is a full solution. The current IMPORT PGDUMP code supports sequences https://github.com/cockroachdb/cockroach/blob/6618d68716ae5a74099843fe768e92df7d50ad19/pkg/ccl/importccl/read_import_pgdump.go#L230 by adding them to a list of descriptors in a resolver https://github.com/cockroachdb/cockroach/blob/6618d68716ae5a74099843fe768e92df7d50ad19/pkg/ccl/importccl/read_import_pgdump.go#L242 and then passing that resolver in https://github.com/cockroachdb/cockroach/blob/6618d68716ae5a74099843fe768e92df7d50ad19/pkg/ccl/importccl/read_import_pgdump.go#L252 and at https://github.com/cockroachdb/cockroach/blob/575ecbcfa0184b3a9cb2b1b3e5f3c12e2bfa2de3/pkg/ccl/importccl/import_stmt.go#L197

I don't think it'd be too hard to do similar SERIAL processing in the IMPORT PGDUMP stuff.

A difficulty may arise when processing CSV, though, since it's never had to make more than 1 descriptor and so lacks some of the plumbing. I'm looking at CSV now to see what would be required for this.

knz commented:
Configuration option is ready and I extended the commit message to explain how it works.

knz commented:
Discussed with @dt @mjibson : ccl import will use the `rowid` strategy always for now. Fixing

knz commented:
New feature has tests now too.

knz commented:
@mjibson I modified `importccl/import_stmt.go` as we discussed. It would be nice if you could review just that.

Also there's another thing. PostgreSQL has the SERIAL pseudo-type always imply NOT NULL (we didn't do that, incorrectly, previously). This patch changes this, so I had to modify the import CSV test. Please have a look at that too. Thanks

mjibson commented on pkg/ccl/importccl/import_stmt.go:
```diff
@@ -191,6 +211,9 @@ func MakeSimpleTableDescriptor(
 	}
 	affected := make(map[sqlbase.ID]*sqlbase.TableDescriptor)
+
+	log.Warningf(ctx, "WOO %q", create.String())
```
Remove.

mjibson commented on pkg/ccl/importccl/csv_internal_test.go:
```diff
@@ -46,11 +46,14 @@ func TestMakeSimpleTableDescriptorErrors(t *testing.T) {
 			stmt:  "create table a (i int, constraint a foreign key (i) references c (id))",
 			error: `this IMPORT format does not support foreign keys`,
 		},
+		{
```
Is this test still accurate? I think this and the next test should be reverted.

mjibson commented on pkg/ccl/importccl/import_stmt.go:
```diff
@@ -168,15 +177,26 @@ func MakeSimpleTableDescriptor(
+			if err != nil {
+				return nil, err
+			}
+			*refTable = tree.MakeUnqualifiedTableName(refTable.TableName)
```
Thanks for fixing this.

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28575)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ccl/importccl/import_stmt.go, line 215 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LJyaZGPTbYeE6z8Z8KR-r8-215:-LK3QC8G5h_C4DcrJqsM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/ccl/importccl/import_stmt.go#L215)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Remove.
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/csv_internal_test.go, line 49 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LJyoU6R6wbcRrxN-izc-r8-49:-LK3QCzoCeazUj8CczdP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/ccl/importccl/csv_internal_test.go#L49)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Is this test still accurate? I think this and the next test should be reverted.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

madelynnblue reviewed (approved):
CCL stuff LGTM.

BramGruneir reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Just a few nits and comments.  

> Reviewed 8 of 11 files at r2, 1 of 1 files at r3, 20 of 20 files at r4, 9 of 29 files at r5, 8 of 11 files at r6, 1 of 1 files at r7, 27 of 27 files at r8, 2 of 2 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-:-LK3UUeu0_rqCRI5aYWE:b-xplgrv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/serial.go, line 116 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3CdLK4mkAOI8xFSyY:-LK3CdLK4mkAOI8xFSyZ:b-18ewn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ed2ef788fcd19511d6d41305f8f4567e3e48c67/pkg/sql/serial.go#L116)):*
> ```Go
> 		return nil, nil, nil, err
> 	}
> 	if res != nil {
> ```

Can you add the first attempt to the loop to?

Might make sense to pull it out into a function.
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 833 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3CqkeAuq2eETpPYBH:-LK3Cqkf6vP8Q4kqrmiz:b592e65) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ed2ef788fcd19511d6d41305f8f4567e3e48c67/pkg/sql/logictest/testdata/logic_test/sequences#L833)):*
> ```
> user root
> 
> # Test virtual sequences.
> ```

subtest please
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 88 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3OhUOBSzck5Rckvav:-LK3OhUOBSzck5Rckvaw:bkdwy9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L88)):*
> ```
> DROP TABLE serials; DROP TABLE smallbig; DROP TABLE serial
> ```

lint: extra lines
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 91 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3PUqCBePxbOcqBbQ4:-LK3PUqD5WTuCRTsC9uM:b-wavvsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L91)):*
> ```
> 
> 
> subtest serial_virtual_sequence
> ```

Please add a test that has to loop 2 times to get a sequence name.
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 182 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3OzEU6R1KlXqHdWvQ:-LK3OzEU6R1KlXqHdWvR:bs4in85) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L182)):*
> ```
> 
> statement ok
> DROP TABLE serials; DROP TABLE smallbig; DROP TABLE serial
> ```

this can be a single drop table statement
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 276 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3POsT3e5TdErugbk3:-LK3POsT3e5TdErugbk4:but1koe) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L276)):*
> ```
> 
> statement ok
> DROP TABLE serials; DROP TABLE smallbig; DROP TABLE serial
> ```

can be a single drop table
___
*[pkg/sql/sqlbase/structured.proto, line 699 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3Lxc1AyCmBQLHQgu5:-LK3Lxc1AyCmBQLHQgu6:b-wolqz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ed2ef788fcd19511d6d41305f8f4567e3e48c67/pkg/sql/sqlbase/structured.proto#L699)):*
> ```protobuf
>     // Start value of the sequence.
>     optional int64 start = 4 [(gogoproto.nullable) = false];
> 	// Whether the sequence is virtual.
> ```

lint: tabs


<!-- Sent from Reviewable.io -->

knz commented:
All tests pass :)

Let me clean this up for you then we'll let the tires kick the (master) road.

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28575)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/serial.go, line 116 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3CdLK4mkAOI8xFSyY:-LK3XbyN72TgG9RNckz8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ed2ef788fcd19511d6d41305f8f4567e3e48c67/pkg/sql/serial.go#L116)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Can you add the first attempt to the loop to?

Might make sense to pull it out into a function.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 833 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3CqkeAuq2eETpPYBH:-LK3WjWBAod1hkTFri-W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ed2ef788fcd19511d6d41305f8f4567e3e48c67/pkg/sql/logictest/testdata/logic_test/sequences#L833)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

subtest please
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 88 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3OhUOBSzck5Rckvav:-LK3V-zV0ooEdbXW5PZ7:b54ire7) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L88)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

lint: extra lines
</blockquote></details>

I liked the little boost to readability of the logic test file. Oh well. Done.
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 91 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3PUqCBePxbOcqBbQ4:-LK3V2KFEH1T1HoD4EKf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L91)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Please add a test that has to loop 2 times to get a sequence name.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 182 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3OzEU6R1KlXqHdWvQ:-LK3Vjlu3DFiUB63Zw_C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L182)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

this can be a single drop table statement
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 276 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3POsT3e5TdErugbk3:-LK3VkW-7h-Eu4YYRb1q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L276)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

can be a single drop table
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.proto, line 699 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3Lxc1AyCmBQLHQgu5:-LK3Vxbb8xO4Ek5JdYm9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ed2ef788fcd19511d6d41305f8f4567e3e48c67/pkg/sql/sqlbase/structured.proto#L699)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

lint: tabs
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

vilterp commented:
Nice work; so speedy! It's not on by default, right? Have to use the `VIRTUAL` keyword? Does it create it automatically if you use a `SERIAL` type?

knz commented:
@vilterp read the release notes :)

jordanlewis commented:
I'm confused - what does that cluster/session setting really do? Does it affect runtime behavior of existing SERIAL datatypes? Or does it only affect what's created when you make a table with SERIAL?

knz commented:
The setting only affects new tables created inside that session.
